Add Data Fetching Guide (#991)
solidjs/solid-docs: src/routes/solid-start/guides/data-fetching.mdx

					<For each={posts()}>{(post) => <li>{post.title}</li>}</For>
				</Suspense>
			</ErrorBoundary>
		</ul>
	);
}
```

## Preloading data

Data fetching can be optimized during user navigation by preloading the data:

1. Export a `route` object with a [`preload`](/solid-router/reference/preload-functions/preload) function.
2. Run your query inside the `preload` function.
3. Use the query as usual in your component.

```tsx {9-11} title="src/routes/index.tsx"
import { ErrorBoundary } from "solid-js";
import { query, createAsync, type RouteDefinition } from "@solidjs/router";

const getPosts = query(async () => {
	const posts = await fetch("https://my-api.com/posts");
	return await posts.json();
}, "posts");

export const route = {
	preload: () => getPosts(),
} satisfies RouteDefinition;

export default function Page() {
	const post = createAsync(() => getPosts());
	return (
		<div>
			<ErrorBoundary fallback={<div>Something went wrong!</div>}>
				<h1>{post().title}</h1>
			</ErrorBoundary>
		</div>
	);
}
```

## Passing parameters to queries

When creating a query that accepts parameters, define your query function to take any number of arguments:

> ```tsx {4} {10} {15} title="src/routes/posts/[id]/index.tsx"
> import { ErrorBoundary } from "solid-js";
> import { query, createAsync, type RouteDefinition } from "@solidjs/router";
> 
> const getPost = query(async (id: string) => {
> 	const post = await fetch(`https://my-api.com/posts/${id}`);
> 	return await post.json();
> }, "post");
> 
> export const route = {
> 	preload: ({ params }) => getPost(params.id),
> } satisfies RouteDefinition;
> 
> export default function Page() {
> 	const postId = 1;
> 	const post = createAsync(() => getPost(postId));
> 	return (
> 		<div>
> 			<ErrorBoundary fallback={<div>Something went wrong!</div>}>
> 				<h1>{post().title}</h1>
> 			</ErrorBoundary>
> 		</div>
> 	);
> }
> ```

Reviewer: Easy to copy/paste snippets from docs is important, and though we're a TypeScript-first resource, it's important not to forget JavaScript users. 

Whenever there are type definitions in a snippet, it's good to provide a tab with a JS-only alternative.

(I had to escape some characters, so I'm not sure this suggestion will really work)

```suggestion
<TabsCodeBlocks>
  <div id="TypeScript">
    ```tsx {4} {10} {15} title="src/routes/posts/[id]/index.tsx"
import { ErrorBoundary } from "solid-js";
import { query, createAsync, type RouteDefinition } from "@solidjs/router";

const getPost = query(async (id: string) => {
	const post = await fetch(`https://my-api.com/posts/${id}`);
	return await post.json();
}, "post");

export const route = {
	preload: ({ params }) => getPost(params.id),
} satisfies RouteDefinition;

export default function Page() {
	const postId = 1;
	const post = createAsync(() => getPost(postId));
	return (
		<div>
			<ErrorBoundary fallback={<div>Something went wrong!</div>}>
				<h1>{post().title}</h1>
			</ErrorBoundary>
		</div>
	);
}
\`\`\`
  </div>
  <div id="JavaScript">
    ```tsx {4} {10} {15} title="src/routes/posts/[id]/index.tsx"
import { ErrorBoundary } from "solid-js";
import { query, createAsync, type RouteDefinition } from "@solidjs/router";

const getPost = query(async (id: string) => {
	const post = await fetch(`https://my-api.com/posts/${id}`);
	return await post.json();
}, "post");

export const route = {
	preload: ({ params }) => getPost(params.id),
};

export default function Page() {
	const postId = 1;
	const post = createAsync(() => getPost(postId));
	return (
		<div>
			<ErrorBoundary fallback={<div>Something went wrong!</div>}>
				<h1>{post().title}</h1>
			</ErrorBoundary>
		</div>
	);
}
\`\`\`
</div>
</TabsCodeBlocks>
```

Author: I added a Javascript version to all examples.

I noticed in other sections of the docs, the tabs are labeled "ts" and "js"  instead of "JavaScript" and "TypeScript". I think the full names are more reasonable. I can update the other sections if you'd like.

	return (
		<ul>
			<ErrorBoundary fallback={<div>Something went wrong!</div>}>
				<Suspense fallback={<div>Loading...</div>}>
					<For each={posts()}>{(post) => <li>{post.title}</li>}</For>
				</Suspense>
			</ErrorBoundary>
		</ul>
	);
}
```

## Preloading data

Data fetching can be optimized during user navigation by preloading the data with the [`preload`](/solid-router/reference/preload-functions/preload) function:

1. Export a `route` object with a `preload` function
2. Run your query inside the `preload` function
3. Use the query as usual in your component

```tsx {9-11} title="src/routes/index.tsx"
import { ErrorBoundary } from "solid-js";
import { query, createAsync, type RouteDefinition } from "@solidjs/router";

const getPosts = query(async () => {
	const posts = await fetch("https://my-api.com/posts");
	return await posts.json();
}, "posts");

export const route = {
	preload: () => getPosts(),
} satisfies RouteDefinition;

export default function Page() {
	const post = createAsync(() => getPosts());
	return (
		<div>
			<ErrorBoundary fallback={<div>Something went wrong!</div>}>
				<h1>{post().title}</h1>
			</ErrorBoundary>
		</div>
	);
}

> ```

Reviewer 2: Should we explain why preloading is even needed?

Reviewer 3: It might be more worthwhile having that in a the other aspects of the docs (like the core parts of "learn" or a "tutorial". This is geared more towards people asking "how to" instead of "why" and i worry if we add the "why" here, it'd be assumed the rest of it should have it as well.

Reviewer 2: that makes sense, we probably need a separate document for explaining the concepts and issues around data fetching + mutation

import { db } from "~/lib/db";

const getPosts = query(async () => {
	"use server";
	return await db.from("posts").select();
}, "posts");

export const route = {
	preload: () => getPosts(),
} satisfies RouteDefinition;

export default function Page() {
	const posts = createAsync(() => getPosts());
	return (
		<ul>
			<ErrorBoundary fallback={<div>Something went wrong!</div>}>
				<For each={posts()}>{(post) => <li>{post.title}</li>}</For>
			</ErrorBoundary>
		</ul>
	);
}
```

## Fetching data on the client

When you only need to fetch data on the client side, you can use the [`createResource`](/reference/basic-reactivity/create-resource) primitive:

```tsx {4-7} title="src/routes/index.tsx"
import { createResource, ErrorBoundary, Suspense, For } from "solid-js";

export default function Page() {
	const [posts] = createResource(async () => {
		const posts = await fetch("https://my-api.com/posts");
		return await posts.json();
	});
	return (
		<ul>
			<ErrorBoundary fallback={<div>Something went wrong!</div>}>
				<Suspense fallback={<div>Loading...</div>}>
					<For each={posts()}>{(post) => <li>{post.title}</li>}</For>
				</Suspense>
			</ErrorBoundary>
		</ul>
	);
}
```

> You can [read more about `createResource` here](/reference/basic-reactivity/create-resource).

Reviewer 2: ~So `createAsync` can still be used to fetch on the client side, and the only reason you'd use `createResource` instead is if you're not using `solid-router`, not if you're fetching on the client. This section seems to confuse those two things together.~

~I'd break this into two separate sections, one that shows fetching on the client with `createAsync` without `"use server"`, and another that shows fetching using `createResource` on both client and server.~

Nevermind, I just noticed that most of your `createAsync` snippets are fetching on the client already. The only correction I have is this section should be called `"Fetching data without Solid Router"` or something like that.

Author: From the [diataxis](https://diataxis.fr/how-to-guides/#how-to-guides-addressed-to-problems) guide that we use as reference:
> How-to guides must be written from the perspective of the user, not of the machinery. A how-to guide represents something that someone needs to get done. It’s defined in other words by the needs of a user. Every how-to guide should answer to a human project, in other words. It should show what the human needs to do, with the tools at hand, to obtain the result they need.
>
> This is in strong contrast to common pattern for how-to guides that often prevails, in which how-to guides are defined by operations that can be performed with a tool or system. The problem with this latter pattern is that it offers little value to the user; it is not addressed to any need the user has. Instead, it’s focused on the tool, on taking the machinery through its motions.

Keeping each section action-oriented, focusing on a practical action that the user might want to do offers better value to the user. If the user wants to learn about a specific tool, it can go to the API Reference.

I believe the documentation should include a section discussing the pros and cons of using Solid.js versus Solid Router APIs for data fetching. Additionally, it would be helpful to clarify why some APIs are part of Solid Router while others belong to Solid.js. However, this page is not aimed at that type of content. Perhaps a tutorial on data fetching would be a more appropriate format for this discussion.

Author: In my initial draft, I included a call-out that gave a little hint:

```
<Callout type="info">
  Note that `query` and `createAsync` are imported from `@solidjs/router`. By default, SolidStart uses [Solid Router](https://docs.solidjs.com/solid-router) under the hood. That means we can take advantage of all Solid Router data fetching primitives in SolidStart.
</Callout>
```

This may slightly disrupt the page's theme, but I think reintroducing a call-out like this could significantly help users navigate the material with greater clarity. At least until we add a comprehensive tutorial.

Reviewer: I came here to say exactly that.
While the guide is straight to the point and practical, I think it leaves loose ends with the `query` and the `createAsync` x `createResource` nuances.

A callout would be nice, but I'd also suggest edits on your proposal above. Solid-Router isn't "under the hood" it's quite on the surface and we actually have work in supporting other client-side routers, so I'd suggest removing that piece. Here's my proposal.

I'd recommend putting such a callout in the very first paragraph.

> Solid-Router has primitives to improve deduplication of resources and requests. Wrapping your fetcher in a `query()` will allow for that. And triggering your fetcher with `createAsync` instead of `createResource` will make sure the reactive system tracks it properly. Except for the deduplication mechanism, `createAsync` and `createResource` are interchangeable.

Author: @atilafassina The problem with your proposal is that it provides no context about the relationship between SolidStart, Solid, and Solid Router. It only scratches the surface of the primitives and capabilities that Solid and Solid Router provide, which I think is outside the scope of this page.

However, I've updated the introduction. What do you think?

Reviewer 3: Going to mark this as resolved for the time being so I can merge this guide. It's probably best at this point to bring it up in a separate issue - @amirhhashemi do you mind making it and summarizing the conversation so we can continue the situation there?

<Callout type="info" title="Advanced Data Handling">
	For advanced features like automatic background re-fetching or infinite queries, you can use [Tanstack Query.](https://tanstack.com/query/latest/docs/framework/solid/overview)

> </Callout>

Reviewer 2: I like the callout, however I still think it would be helpful to show an example snippet here that shows using tanstack query with `"use server"`. Would love to get some more thoughts on this.

Reviewer 3: I almost wonder if we need a tutorial geared towards advanced data fetching. Like something that goes into the weeds of the more complicated parts people may run into?

Author: I think showcasing the usage of Tanstack Query with `"use server"` is interesting enough to include in this guide. But, my concern is that it might go too far. If we want to cover Tanstack Query properly, we should add:

1. A basic example of `createQuery` (with and without `"use server"`)
2. A section about preloading data
3. A section about error handling and loading UI
4. A few examples of advanced things that Tanstack Query can do and SolidStart can't

It's too much for this page. I think it's better to leave it to the user to go through the Tanstack Query docs and figure things out. If there's any ambiguity regarding the use of `"use server"` with Tanstack Query, we should consider updating other sections of the documentation for clarity.

Reviewer 3: I'd maybe say this is more something for the Tanstack docs than our own. Let me bring this up to some people within the team, as well as Tanner, and see if we can better hammer this out.

Reviewer 3: There's some stuff incoming for Tanstack - going to resolve this conversation for the time being but we will have to revisit this in the future.

---
title: "Data Fetching"
---

This guide will walk you through the basics of data fetching in SolidStart, providing practical examples and best practices.

Here's a minimal example of data fetching in SolidStart:

```tsx title="src/routes/index.tsx"
import { For } from "solid-js";
import { query, createAsync } from "@solidjs/router";

const getPosts = query(async () => {
	const posts = await fetch("https://my-api.com/blog");
	return await posts.json();
}, "posts");

export default function Page() {
	const posts = createAsync(() => getPosts());
	return (
		<ul>
			<For each={posts()}>{(post) => <li>{post.title}</li>}</For>
		</ul>
	);
}
```

> In this example, we created a [`query`](https://docs.solidjs.com/solid-router/reference/data-apis/query) and used the [`createAsync`](https://docs.solidjs.com/solid-router/reference/data-apis/create-async) primitive to access the query data inside our component.

Reviewer 3: ```suggestion
In this example, a [`query`](/solid-router/reference/data-apis/query) is created. In order to access it's data within the component, the [`createAsync`](/solid-router/reference/data-apis/create-async) primitive was used.
```

## Reference

- Solid Router [`query`](https://docs.solidjs.com/solid-router/reference/data-apis/query)
- Solid Router [`createAsync`](https://docs.solidjs.com/solid-router/reference/data-apis/create-async)
- Solid Router [`preload`](https://docs.solidjs.com/solid-router/reference/preload-functions/preload)
- Solid [`createResource`](https://docs.solidjs.com/reference/basic-reactivity/create-resource)
- Solid [`<Suspense>`](https://docs.solidjs.com/reference/components/suspense)

> - Solid [`<ErrorBoundary>`](https://docs.solidjs.com/reference/components/error-boundary)

Reviewer 3: I'm not entirely sure this section is necessary since each part is linked out when used

## Showing loading UI

> To show a loading UI while data is loading, you can use [Solid `<Suspense>`](https://docs.solidjs.com/reference/components/suspense):

Reviewer 3: ```suggestion
To display a loading UI while data is being fetched, you can use [Solid `<Suspense>`](/reference/components/suspense):
```

1. Import `Suspense` from `solid-js`
2. Wrap the data rendering in `<Suspense>`
3. Provide a fallback component that will be displayed while the data is loading using the `fallback` prop

```tsx {1} {13} {15} title="src/routes/index.tsx"
import { Suspense, For } from "solid-js";
import { query, createAsync } from "@solidjs/router";

const getPosts = query(async () => {
	const posts = await fetch("https://my-api.com/blog");
	return await posts.json();
}, "posts");

export default function Page() {
	const posts = createAsync(() => getPosts());
	return (
		<ul>
			<Suspense fallback={<div>Loading...</div>}>
				<For each={posts()}>{(post) => <li>{post.title}</li>}</For>
			</Suspense>
		</ul>
	);
}
```

## Handling errors

> To show a fallback UI in case data fetching fails, you can use [Solid `<ErrorBoundary>`](https://docs.solidjs.com/reference/components/error-boundary):

Reviewer 3: ```suggestion
If the data fetching fails, a fallback UI can be displayed using an [`<ErrorBoundary>`](/reference/components/error-boundary):
```

1. Import `ErrorBoundary` from `solid-js`.
2. Wrap the data rendering in `<ErrorBoundary>`
3. Provide a fallback component that will be displayed if an error occurs using the `fallback` prop.

```tsx {1} {13} {17} title="src/routes/index.tsx"
import { ErrorBoundary, Suspense, For } from "solid-js";
import { query, createAsync } from "@solidjs/router";

const getPosts = query(async () => {
	const posts = await fetch("https://my-api.com/blog");
	return await posts.json();
}, "posts");

export default function Page() {
	const posts = createAsync(() => getPosts());
	return (
		<ul>
			<ErrorBoundary fallback={<div>Something went wrong!</div>}>
				<Suspense fallback={<div>Loading...</div>}>
					<For each={posts()}>{(post) => <li>{post.title}</li>}</For>
				</Suspense>
			</ErrorBoundary>
		</ul>
	);
}
```

## Preloading data

> To optimize data fetching during user navigation, you can use [`preload`](https://docs.solidjs.com/solid-router/reference/preload-functions/preload) to preload a query:

Reviewer 3: ```suggestion
Data fetching can be optimized during user navigation through preloading the data with the [`preload`](/solid-router/reference/preload-functions/preload) function: 
```

1. Export a `route` object with a `preload` function
2. Run your query inside the `preload` function
3. Use the query as usual in your component

```tsx {2} {9-11} title="src/routes/index.tsx"
import { ErrorBoundary } from "solid-js";
import { query, createAsync, type RouteDefinition } from "@solidjs/router";

const getPosts = query(async () => {
	const posts = await fetch("https://my-api.com/blog");
	return await posts.json();
}, "posts");

export const route = {
	preload: () => getPosts(),
} satisfies RouteDefinition;

export default function Page() {
	const post = createAsync(() => getPosts());
	return (
		<div>
			<ErrorBoundary fallback={<div>Something went wrong!</div>}>
				<h1>{post().title}</h1>
			</ErrorBoundary>
		</div>
	);
}
```

## Passing parameters to queries

> To create a query that accepts parameters, you can pass any number of arguments to your query function:

Reviewer 3: ```suggestion
When creating a query that accepts parameters, you can define your query function to take any number of arguments:
```

```tsx {9-10} {15} {19-20} title="src/routes/posts/[id]/index.tsx"
import { ErrorBoundary } from "solid-js";
import {
	query,
	createAsync,
	useParams,
	type RouteDefinition,
} from "@solidjs/router";

const getPost = query(async (id: string) => {
	const post = await fetch(`https://my-api.com/blog/posts/${id}`);
	return await post.json();
}, "post");

export const route = {
	preload: ({ params }) => getPost(params.id),
} satisfies RouteDefinition;

export default function Page() {
	const params = useParams();
	const post = createAsync(() => getPost(params.id));
	return (
		<div>
			<ErrorBoundary fallback={<div>Something went wrong!</div>}>
				<h1>{post().title}</h1>
			</ErrorBoundary>
		</div>
	);
}
```

> In this example, we used [`useParams`](https://docs.solidjs.com/solid-router/reference/primitives/use-params) to get the `id` path parameter and passed it to the `getPost` query to get the data for that specific post.

Reviewer 3: ```suggestion
In this example, [`useParams`](/solid-router/reference/primitives/use-params) was used to access the `id` path parameter. You can then pass it to the `getPost` query to get the data for that specific post.
```

## Using database or an ORM

> To directly use your database or ORM in a query, you should make the query server-only so your database credentials are not exposed to the client. You can do that by adding [`"use server"`](https://docs.solidjs.com/solid-start/reference/server/use-server) to the first line of your query function:

Reviewer 3: ```suggestion
To safely interact with your database or ORM in a query, ensure the query is server-only to prevent exposing your database credentials to the client. You can do this by adding [`"use server"`](/solid-start/reference/server/use-server) as the first line of your query function:
```

```tsx {3} {6-7} title="src/routes/index.tsx"
import { For, ErrorBoundary } from "solid-js";
import { query, createAsync, type RouteDefinition } from "@solidjs/router";
import { db } from "~/lib/db";

const getPosts = query(async () => {
	"use server";
	return await db.from("posts").select();
}, "posts");

export const route = {
	preload: () => getPosts(),
} satisfies RouteDefinition;

export default function Page() {
	const posts = createAsync(() => getPosts());
	return (
		<ul>
			<ErrorBoundary fallback={<div>Something went wrong!</div>}>
				<For each={posts()}>{(post) => <li>{post.title}</li>}</For>
			</ErrorBoundary>
		</ul>
	);
}
```

## Fetching data on the client

> To fetch data only on the client side, you can use the [`createResource`](https://docs.solidjs.com/reference/basic-reactivity/create-resource) primitive:

Reviewer 3: ```suggestion
When you only need to fetch data on the client side, you can use the [`createResource`](/reference/basic-reactivity/create-resource) primitive:
```

```tsx {1} {4-7} title="src/routes/index.tsx"
import { createResource, ErrorBoundary, Suspense, For } from "solid-js";

export default function Page() {
	const [posts] = createResource(async () => {
		const posts = await fetch("https://my-api.com/blog");
		return await posts.json();
	});
	return (
		<ul>
			<ErrorBoundary fallback={<div>Something went wrong!</div>}>
				<Suspense fallback={<div>Loading...</div>}>
					<For each={posts()}>{(post) => <li>{post.title}</li>}</For>
				</Suspense>
			</ErrorBoundary>
		</ul>
	);
}
```

> Read more in the [`createResource`](https://docs.solidjs.com/reference/basic-reactivity/create-resource) docs.

Reviewer 3: ```suggestion
You can [read more about `createResource` here](/reference/basic-reactivity/create-resource) .
```

> <Callout type="info">

Reviewer 3: ```suggestion
<Callout type="info" title="Advanced Data Handling">
```

>     [Tanstack Query](https://tanstack.com/query/latest/docs/framework/solid/overview) can be used for more advanced features such as automatic background re-fetching or infinite queries.

Reviewer 3: ```suggestion
    For advanced features like automatic background re-fetching or infinite queries, you can use [Tanstack Query.](https://tanstack.com/query/latest/docs/framework/solid/overview)
```
